woocommerce/google-listings-and-ads
# Sync products for all target countries

### Changes proposed in this Pull Request:

Part of #14.

This PR modifies the ProductSyncer to take into account the target audience countries the user has set when submitting products. This means that each product is now submitted for all of the countries that the user has set as their target audience in the configurations/onboarding step.

Google assigns each product+targetCountry combination a unique ID, meaning that we have a one-to-many relationship between WooCommerce product IDs and Google IDs. The IDs are stored as an array in the `_wc_gla_google_ids` meta.

When a user tries to delete the product from Google MC (i.e. not sync it anymore), we might send several batches of delete requests to Google with all of the Google IDs.

In case of any errors while syncing/deleting a product there are actions that are triggered and schedule a second update/delete job with just the errored products. We only retry the update/delete if the error is of type "Internal Error", which is basically a Google backend error. This error has appeared several times during my testing especially once the number of products that are submitted is large (the batch size is still way less than the recommended size of 1000 so the error seems to be unrelated to the request).

In order to avoid any timeouts, the size of each batch when submitting products is set based on the number of target audience countries set by the user.

### Detailed test instructions:

1. Make sure you have the `gla_mc_setup_completed_at` option set to a timestamp in `wp_options` table. (i.e. your merchant center set up is completed)
1. Go to Marketing > Google and select some target countries or "All Countries" from the "Choose your audience" step
1. ~~Go to the Connection Test page and initiate an async product syncing job.~~ A job called `gla/jobs/refresh_synced_products/create_batch` should be scheduled to resubmit the products to Google MC
1. Wait for the job to complete. There might be several jobs and they may take a while to complete if you have many products/variations and/or many target countries.
1. Check Google Merchant center and under "Feeds" make sure that all target countries have an equal number of submitted products
1. If they don't have an equal number of products per country, it might be due to some syncing errors. If the errors are of type "Internal Error" there is a job implemented that tries to resync those products until they get synced. However, if the error is any other type, the products may not be synced for some countries (e.g. validation errors specific to some countries).

Note that the synchronous mode of syncing might not work properly if your number of products or the number of target countries is large. You may hit a time-out error if you use the synchronous mode.

### To do:

- [x] Make sure the delete job deletes all products from all countries (+ optimize it)
- [x] Refresh the product feed once the target country changes (i.e. remove it from a country if the user deselects it as a target)
- [x] Fix the RefreshProducts job queued multiple times on audience change
